volumes.cpp

```diff
@@ -1312,7 +1331,249 @@ SurfaceVolume& makeSurface( const SurfaceCard* card, VolumeCache* v){
+    return rppFacet( args, facet );
+  }
+  else if( mnemonic == "hex" || mnemonic == "rhp" ){
+    return hexFacet( args, facet );
+  }
+  else if( mnemonic == "rec" ){
+    return recFacet( args, facet );
+  }
+  throw std::runtime_error( mnemonic + " does not have macrobody facet support at this time." );
+}    
+
+
+//This function is used to find D in the planar equation Ax + By + Cz - D = 0 when given two vectors.
+int planePoint( std::vector< double > args, int i, bool end ){
+  if( end ){
+    return args.at(i)*( args.at(0) + args.at(i) ) + args.at(i + 1)*( args.at(1) + args.at(i + 1) ) + args.at(i + 2)*( args.at(2) + args.at(i + 2) );
+  }
+  else{
+    return args.at(i)*args.at(0) + args.at(i + 1)*args.at(1) + args.at(i + 2)*args.at(2);
+  }
+}
+
+
+SurfaceVolume* rccFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
+    //cylinder surface
+    return new RccVolume( Vector3d(args), Vector3d(args,3), args.at(6), true );
+  }
+  else if( facet == 2 ){
+    //plane at end of vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, true );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 3 ){
+    //plane at start of vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, false );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "rcc only has 3 facets" );
+  }
+}
+
+SurfaceVolume* boxFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
+    //end of first vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, true );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 2 ){
+    //beginning of first vector
+    Vector3d v( args.at(3), args.at(4), args.at(5) );
+    double D = planePoint( args, 3, false );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 3 ){
+    //end of second vector
+    Vector3d v( args.at(6), args.at(7), args.at(8) );
+    double D = planePoint( args, 6, true );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 4 ){
+    //beginning of second vector
+    Vector3d v( args.at(6), args.at(7), args.at(8) );
+    double D = planePoint( args, 6, false );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 5 ){
+    //end of third vector
+    Vector3d v( args.at(9), args.at(10), args.at(11) );
+    double D = planePoint( args, 9, true );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else if( facet == 6 ){
+    //beginning of third vector
+    Vector3d v( args.at(9), args.at(10), args.at(11) );
+    double D = planePoint( args, 9, false );
+    return new PlaneSurface( v, D/v.length() );
+  }
+  else{
+    throw std::runtime_error( "box only has 6 facets");
+  }
+}
+
+SurfaceVolume* rppFacet( const std::vector< double > args, int facet ){
+  if( facet == 1 ){
```

Comment:
ibid: there is a simple mathematical way to define the vector and args here and then call a single line of code